File: ios/TextRecognitionPlugin.swift
//
//  TextRecognitionPlugin.swift
//  react-native-vision-camera-ml-kit
//

import Foundation
import VisionCamera
import MLKitVision
import MLKitCommon
import MLKitTextRecognition
import MLKitTextRecognitionChinese
import MLKitTextRecognitionDevanagari
import MLKitTextRecognitionJapanese
import MLKitTextRecognitionKorean

@objc(TextRecognitionPlugin)
public class TextRecognitionPlugin: FrameProcessorPlugin {

    private var textRecognizer: TextRecognizer!
    private let processingLock = NSLock()
    private var isProcessing = false

    public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
        super.init(proxy: proxy, options: options)

        let language = (options["language"] as? String ?? "latin").lowercased()
        Logger.info("Initializing text recognition with language: \(language)")

        switch language {
        case "chinese":
            textRecognizer = TextRecognizer.textRecognizer(options: ChineseTextRecognizerOptions())
        case "devanagari":
            textRecognizer = TextRecognizer.textRecognizer(options: DevanagariTextRecognizerOptions())
        case "japanese":
            textRecognizer = TextRecognizer.textRecognizer(options: JapaneseTextRecognizerOptions())
        case "korean":
            textRecognizer = TextRecognizer.textRecognizer(options: KoreanTextRecognizerOptions())
        case "latin", "default":
            textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
        default:
            Logger.warn("Unknown language '\(language)', defaulting to Latin")
            textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
        }

        Logger.info("Text recognition initialized successfully")
    }

    deinit {
        // Clean up ML Kit resources when plugin is deallocated
        // Swift ARC will handle the deallocation, but we log for debugging
        Logger.debug("TextRecognitionPlugin deallocating - ML Kit recognizer resources will be freed")
        // Note: ML Kit resources are automatically freed by ARC when textRecognizer is deallocated
    }

    public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
        // Skip frame if previous processing is still in progress
        processingLock.lock()
        if isProcessing {
            processingLock.unlock()
            Logger.debug("Skipping frame - previous processing still in progress")
            return nil
        }
        isProcessing = true
        processingLock.unlock()

        // Use Vision Camera's reference counting to keep frame alive during processing
        frame.incrementRefCount()

        let startTime = Date()

        defer {
            // Release frame reference and reset processing flag
            frame.decrementRefCount()
            processingLock.lock()
            isProcessing = false
            processingLock.unlock()
        }

        do {
            let orientation = frame.orientation

            let visionImage = VisionImage(buffer: frame.buffer)
            visionImage.orientation = getOrientation(orientation: orientation)

            Logger.debug("Processing frame: \(frame.width)x\(frame.height), orientation: \(orientation.rawValue)")

            // Process synchronously
            let text = try textRecognizer.results(in: visionImage)

            let processingTime = Int64(Date().timeIntervalSince(startTime) * 1000)
            Logger.performance("Text recognition processing", durationMs: processingTime)

            if text.text.isEmpty {
                Logger.debug("No text detected in frame")
                return nil
            }

            Logger.debug("Text detected: \(text.text.count) characters, \(text.blocks.count) blocks")

            return [
                "text": text.text,
                "blocks": processBlocks(text.blocks)
            ]

        } catch {
            let processingTime = Int64(Date().timeIntervalSince(startTime) * 1000)
            Logger.error("Error during text recognition: \(error.localizedDescription)")
            Logger.performance("Text recognition processing (error)", durationMs: processingTime)
            return nil
        }
    }

    // MARK: - Orientation Mapping

    private func getOrientation(orientation: UIImage.Orientation) -> UIImage.Orientation {
        switch orientation {
        case .up:
            return .up
        case .left:
            return .right  // Swap left and right
        case .down:
            return .down
        case .right:
            return .left   // Swap left and right
        default:
            return .up
        }
    }

    // MARK: - Helper Methods

    private func processBlocks(_ blocks: [TextBlock]) -> [[String: Any]] {
        return blocks.map { block in
            var blockDict: [String: Any] = [
                "text": block.text,
                "frame": processRect(block.frame),
                "cornerPoints": processCornerPoints(block.cornerPoints),
                "lines": processLines(block.lines)
            ]

            if let lang = block.recognizedLanguages.first?.languageCode {
                blockDict["recognizedLanguage"] = lang
            }

            return blockDict
        }
    }

    private func processLines(_ lines: [TextLine]) -> [[String: Any]] {
        return lines.map { line in
            var lineDict: [String: Any] = [
                "text": line.text,
                "frame": processRect(line.frame),
                "cornerPoints": processCornerPoints(line.cornerPoints),
                "elements": processElements(line.elements)
            ]

            if let lang = line.recognizedLanguages.first?.languageCode {
                lineDict["recognizedLanguage"] = lang
            }

            return lineDict
        }
    }

    private func processElements(_ elements: [TextElement]) -> [[String: Any]] {
        return elements.map { element in
            var elementDict: [String: Any] = [
                "text": element.text,
                "frame": processRect(element.frame),
                "cornerPoints": processCornerPoints(element.cornerPoints)
            ]

            if let lang = element.recognizedLanguages.first?.languageCode {
                elementDict["recognizedLanguage"] = lang
            }

            return elementDict
        }
    }

    private func processRect(_ rect: CGRect) -> [String: CGFloat] {
        return [
            "x": rect.midX,
            "y": rect.midY,
            "width": rect.width,
            "height": rect.height
        ]
    }

    private func processCornerPoints(_ cornerPoints: [NSValue]) -> [[String: Int]] {
        return cornerPoints.compactMap { $0.cgPointValue }.map { point in
            ["x": Int(point.x), "y": Int(point.y)]
        }
    }
}

File: android/src/main/java/com/rnvisioncameramlkit/TextRecognitionPlugin.kt
package com.rnvisioncameramlkit

import android.graphics.Point
import android.graphics.Rect
import android.media.Image
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.TextRecognizerOptionsInterface
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import com.rnvisioncameramlkit.utils.Logger
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Text Recognition v2 Frame Processor Plugin
 *
 * Performs on-device text recognition using Google ML Kit
 * Supports multiple scripts: Latin, Chinese, Devanagari, Japanese, Korean
 */
class TextRecognitionPlugin(
    proxy: VisionCameraProxy,
    options: Map<String, Any>?
) : FrameProcessorPlugin() {

    private var recognizer: TextRecognizer
    private val isProcessing = AtomicBoolean(false)

    init {
        val language = options?.get("language")?.toString() ?: "latin"
        Logger.info("Initializing text recognition with language: $language")

        val recognizerOptions: TextRecognizerOptionsInterface = when (language.lowercase()) {
            "chinese" -> ChineseTextRecognizerOptions.Builder().build()
            "devanagari" -> DevanagariTextRecognizerOptions.Builder().build()
            "japanese" -> JapaneseTextRecognizerOptions.Builder().build()
            "korean" -> KoreanTextRecognizerOptions.Builder().build()
            "latin", "default" -> TextRecognizerOptions.Builder().build()
            else -> {
                Logger.warn("Unknown language '$language', defaulting to Latin")
                TextRecognizerOptions.Builder().build()
            }
        }

        recognizer = TextRecognition.getClient(recognizerOptions)
        Logger.info("Text recognition initialized successfully")
    }

    /**
     * Cleanup resources when plugin is destroyed
     * Called automatically by finalize() when the plugin is garbage collected,
     * or can be called manually to release resources earlier.
     */
    fun cleanup() {
        try {
            // Close ML Kit recognizer to release native resources
            recognizer.close()
            Logger.debug("Text recognizer resources cleaned up successfully")
        } catch (e: Exception) {
            Logger.error("Error cleaning up text recognizer resources", e)
        }
    }

    /**
     * Finalizer to ensure cleanup happens when plugin is garbage collected
     * This prevents memory leaks of ML Kit native resources (models, GPU memory)
     */
    @Suppress("DEPRECATION")
    protected fun finalize() {
        cleanup()
    }

    override fun callback(frame: Frame, arguments: Map<String, Any>?): Any? {
        // Skip frame if previous processing is still in progress
        if (!isProcessing.compareAndSet(false, true)) {
            Logger.debug("Skipping frame - previous processing still in progress")
            return null
        }

        val startTime = System.currentTimeMillis()

        try {
            val mediaImage: Image = frame.image
            val image = InputImage.fromMediaImage(
                mediaImage,
                frame.imageProxy.imageInfo.rotationDegrees
            )

            Logger.debug("Processing frame: ${frame.width}x${frame.height}, rotation: ${frame.imageProxy.imageInfo.rotationDegrees}")

            val task: Task<Text> = recognizer.process(image)
            val text: Text = Tasks.await(task)

            val processingTime = System.currentTimeMillis() - startTime
            Logger.performance("Text recognition processing", processingTime)

            if (text.text.isEmpty()) {
                Logger.debug("No text detected in frame")
                return null
            }

            Logger.debug("Text detected: ${text.text.length} characters, ${text.textBlocks.size} blocks")

            val result = WritableNativeMap().apply {
                putString("text", text.text)
                putArray("blocks", processBlocks(text.textBlocks))
            }

            return result.toHashMap()

        } catch (e: Exception) {
            val processingTime = System.currentTimeMillis() - startTime
            Logger.error("Error during text recognition", e)
            Logger.performance("Text recognition processing (error)", processingTime)
            return null
        } finally {
            isProcessing.set(false)
        }
    }

    companion object {
        /**
         * Process text blocks into React Native compatible format
         */
        private fun processBlocks(blocks: List<Text.TextBlock>): WritableNativeArray {
            val blockArray = WritableNativeArray()

            for (block in blocks) {
                val blockMap = WritableNativeMap().apply {
                    putString("text", block.text)
                    putMap("frame", processRect(block.boundingBox))
                    putArray("cornerPoints", processCornerPoints(block.cornerPoints))
                    putArray("lines", processLines(block.lines))

                    // Add language if recognized
                    block.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }

                    // Add confidence if available (ML Kit doesn't provide this for v2, but keeping for future)
                    // putDouble("confidence", block.confidence?.toDouble() ?: 0.0)
                }
                blockArray.pushMap(blockMap)
            }

            return blockArray
        }

        /**
         * Process text lines into React Native compatible format
         */
        private fun processLines(lines: List<Text.Line>): WritableNativeArray {
            val lineArray = WritableNativeArray()

            for (line in lines) {
                val lineMap = WritableNativeMap().apply {
                    putString("text", line.text)
                    putMap("frame", processRect(line.boundingBox))
                    putArray("cornerPoints", processCornerPoints(line.cornerPoints))
                    putArray("elements", processElements(line.elements))

                    line.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }
                }
                lineArray.pushMap(lineMap)
            }

            return lineArray
        }

        /**
         * Process text elements (words) into React Native compatible format
         */
        private fun processElements(elements: List<Text.Element>): WritableNativeArray {
            val elementArray = WritableNativeArray()

            for (element in elements) {
                val elementMap = WritableNativeMap().apply {
                    putString("text", element.text)
                    putMap("frame", processRect(element.boundingBox))
                    putArray("cornerPoints", processCornerPoints(element.cornerPoints))
                    putArray("symbols", processSymbols(element.symbols))

                    element.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }
                }
                elementArray.pushMap(elementMap)
            }

            return elementArray
        }

        /**
         * Process text symbols (characters) into React Native compatible format
         */
        private fun processSymbols(symbols: List<Text.Symbol>): WritableNativeArray {
            val symbolArray = WritableNativeArray()

            for (symbol in symbols) {
                val symbolMap = WritableNativeMap().apply {
                    putString("text", symbol.text)
                    putMap("frame", processRect(symbol.boundingBox))
                    putArray("cornerPoints", processCornerPoints(symbol.cornerPoints))

                    symbol.recognizedLanguage?.let { lang ->
                        putString("recognizedLanguage", lang)
                    }
                }
                symbolArray.pushMap(symbolMap)
            }

            return symbolArray
        }

        /**
         * Convert Android Rect to React Native format
         */
        private fun processRect(boundingBox: Rect?): WritableNativeMap {
            val rectMap = WritableNativeMap()

            boundingBox?.let { box ->
                rectMap.putDouble("x", box.exactCenterX().toDouble())
                rectMap.putDouble("y", box.exactCenterY().toDouble())
                rectMap.putInt("width", box.width())
                rectMap.putInt("height", box.height())
            }

            return rectMap
        }

        /**
         * Convert Android corner points to React Native format
         */
        private fun processCornerPoints(cornerPoints: Array<Point>?): WritableNativeArray {
            val pointsArray = WritableNativeArray()

            cornerPoints?.forEach { point ->
                val pointMap = WritableNativeMap().apply {
                    putInt("x", point.x)
                    putInt("y", point.y)
                }
                pointsArray.pushMap(pointMap)
            }

            return pointsArray
        }
    }
}
